File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/D-Ход_конем_2/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(NM)
    Memory Complexity O(NM)
    */
    int n, m;
    std::scanf("%d %d", &n, &m);
    std::vector<std::vector<long long>> dp (n, std::vector<long long>(m, 0));
    dp[0][0] = 1;
    std::vector<std::pair<int, int>> moves = {
        {-1, -2}, {-2, -1}, {1, -2}, {-2, 1}
    };
    int mainR = 0;
    while (mainR < n) {
        int r = mainR;
        int c = 0;
        while (r >= 0 && c < m) {
            for (std::pair<int, int>& move : moves) {
                int nextR = r + move.first;
                int nextC = c + move.second;
                if (nextR >= 0 && nextR < n && nextC >= 0 and nextC < m) {
                    dp[r][c] += dp[nextR][nextC];
                }
            }
            --r;
            ++c;
        }
        ++mainR;
    }
    int mainC = 1;
    while (mainC < m) {
        int r = n - 1;
        int c = mainC;
        while (r >= 0 && c < m) {
            for (std::pair<int, int>& move : moves) {
                int nextR = r + move.first;
                int nextC = c + move.second;
                if (nextR >= 0 && nextR < n && nextC >= 0 and nextC < m) {
                    dp[r][c] += dp[nextR][nextC];
                }
            }
            --r;
            ++c;
        }
        ++mainC;
    }
    std::printf("%lld\n", dp[n - 1][m - 1]);
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/D-Ход_конем_2/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NM)
    Memory Complexity O(NM)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    dp: list[list[int]] = [[0] * m for _ in range(n)]
    moves: list[tuple[int, int]] = [
        (-1, -2), (-2, -1), (1, -2), (-2, 1)
    ]
    dp[0][0] = 1
    main_r: int = 0
    while main_r < n:
        r: int = main_r
        c: int = 0
        while r >= 0 and c < m:
            for x, y in moves:
                next_r: int = r + x
                next_c: int = c + y
                if (0 <= next_r < n) and (0 <= next_c < m):
                    dp[r][c] += dp[next_r][next_c]
            r -= 1 
            c += 1
        main_r += 1

    main_c: int = 1
    while main_c < m:
        r: int = n - 1
        c: int = main_c
        while r >= 0 and c < m:
            for x, y in moves:
                next_r: int = r + x
                next_c: int = c + y
                if (0 <= next_r < n) and (0 <= next_c < m):
                    dp[r][c] += dp[next_r][next_c]
            r -= 1 
            c += 1
        main_c += 1
    sys.stdout.write('{}\n'.format(dp[n - 1][m - 1]))


if __name__ == '__main__':
    solution()